### ISSUE TITLE
Python asserts if deadline larger than 0x3c000000

### ISSUE THREAD

**Issue by bryanschwerer** [grpc/grpc#5827]
When making a GRPC call with a value larger than 0x3c000000

E0318 09:28:32.923703185   31463 time.c:185]                 assertion failed: b.clock_type == GPR_TIMESPAN
Aborted

Note:  If this is some bit pattern thing, it may not be a larger value, my experience was  0x3c000000 passed,  0x3c000001 did not.

**Comment by ctiller** [comment]
Looks like a misuse of API

**Comment by soltanmm** [comment]
@bryanschwerer Is the hex value you provided a Python `int` or a `float`?

**Comment by bryanschwerer** [comment]
Int.  The reason I used hex was that it made the binary search for the offending value easier.

**Comment by soltanmm** [comment]
@ctiller I'm doubtful it's misuse of the API (or that if there is misuse it's secondary to a problem that should be addressed in core *anyways*); the stack backtrace on the abortion shows that it's inside the HTTP/2 HPACK parser:

```
#0  0x00007ffff782dcc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff78310d8 in __GI_abort () at abort.c:89
#2  0x00007ffff1b58986 in gpr_time_add (a=..., b=...) at src/core/support/time.c:185
#3  0x00007ffff1b7d501 in on_initial_header (tp=0x7fffe003dd70, md=0x7fffd8006050) at src/core/transport/chttp2/parsing.c:611
#4  0x00007ffff1b7b556 in on_hdr (p=p@entry=0x7fffe003de60, md=0x7fffd8006050, add_to_table=add_to_table@entry=1) at src/core/transport/chttp2/hpack_parser.c:639
#5  0x00007ffff1b7bf37 in finish_lithdr_incidx_v (p=0x7fffe003de60, cur=0x7fffe003e907 "", end=0x7fffe003e907 "") at src/core/transport/chttp2/hpack_parser.c:768
#6  0x00007ffff1b7c5a2 in parse_next (end=0x7fffe003e907 "", cur=0x7fffe003e907 "", p=0x7fffe003de60) at src/core/transport/chttp2/hpack_parser.c:654
#7  parse_string (p=0x7fffe003de60, cur=<optimized out>, end=0x7fffe003e907 "") at src/core/transport/chttp2/hpack_parser.c:1293
#8  0x00007ffff1b7c5a2 in parse_next (end=0x7fffe003e907 "", cur=0x7fffe003e8fb "\v2006632961S", p=0x7fffe003de60) at src/core/transport/chttp2/hpack_parser.c:654
#9  parse_string (p=0x7fffe003de60, cur=<optimized out>, end=0x7fffe003e907 "") at src/core/transport/chttp2/hpack_parser.c:1293
#10 0x00007ffff1b7bf64 in parse_begin (end=0x7fffe003e907 "", cur=0x7fffe003e8ed "@\fgrpc-timeout\v2006632961S", p=0x7fffe003de60) at src/core/transport/chttp2/hpack_parser.c:666
#11 finish_lithdr_incidx_v (p=0x7fffe003de60, cur=0x7fffe003e8ed "@\fgrpc-timeout\v2006632961S", end=0x7fffe003e907 "") at src/core/transport/chttp2/hpack_parser.c:771
#12 0x00007ffff1b7c5a2 in parse_next (end=0x7fffe003e907 "", cur=0x7fffe003e8ed "@\fgrpc-timeout\v2006632961S", p=0x7fffe003de60) at src/core/transport/chttp2/hpack_parser.c:654
#13 parse_string (p=0x7fffe003de60, cur=<optimized out>, end=0x7fffe003e907 "") at src/core/transport/chttp2/hpack_parser.c:1293
#14 0x00007ffff1b7c5a2 in parse_next (end=0x7fffe003e907 "", cur=0x7fffe003e8bb "1Python-gRPC-0.14.0.dev0 grpc-c/0.14.0-dev (linux)@\fgrpc-timeout\v2006632961S", p=0x7fffe003de60)
    at src/core/transport/chttp2/hpack_parser.c:654
#15 parse_string (p=0x7fffe003de60, cur=<optimized out>, end=0x7fffe003e907 "") at src/core/transport/chttp2/hpack_parser.c:1293
#16 0x00007ffff1b7bf64 in parse_begin (end=0x7fffe003e907 "", cur=0x7fffe003e8af "@\nuser-agent1Python-gRPC-0.14.0.dev0 grpc-c/0.14.0-dev (linux)@\fgrpc-timeout\v2006632961S", p=0x7fffe003de60)
    at src/core/transport/chttp2/hpack_parser.c:666
#17 finish_lithdr_incidx_v (p=0x7fffe003de60, cur=0x7fffe003e8af "@\nuser-agent1Python-gRPC-0.14.0.dev0 grpc-c/0.14.0-dev (linux)@\fgrpc-timeout\v2006632961S", end=0x7fffe003e907 "")
    at src/core/transport/chttp2/hpack_parser.c:771
#18 0x00007ffff1b7c5a2 in parse_next (end=0x7fffe003e907 "", cur=0x7fffe003e8af "@\nuser-agent1Python-gRPC-0.14.0.dev0 grpc-c/0.14.0-dev (linux)@\fgrpc-timeout\v2006632961S", p=0x7fffe003de60)
    at src/core/transport/chttp2/hpack_parser.c:654
#19 parse_string (p=0x7fffe003de60, cur=<optimized out>, end=0x7fffe003e907 "") at src/core/transport/chttp2/hpack_parser.c:1293
#20 0x00007ffff1b7c5a2 in parse_next (end=0x7fffe003e907 "", cur=0x7fffe003e89e "\020application/grpc@\nuser-agent1Python-gRPC-0.14.0.dev0 grpc-c/0.14.0-dev (linux)@\fgrpc-timeout\v2006632961S", 
    p=0x7fffe003de60) at src/core/transport/chttp2/hpack_parser.c:654
...
```

**Comment by soltanmm** [comment]
To the best of my knowledge that's outside the realm of Python's control, and if it isn't then what leads to that control *may* be a Bad Thing™. I'll keep poking, though. I'm confuzzed by this one (and perhaps fuzzing may be an answer).

**Comment by soltanmm** [comment]
@bryanschwerer When you encountered this bug, was the abortion on the server side or the client side?

**Comment by bryanschwerer** [comment]
Server side.

**Comment by soltanmm** [comment]
So we don't forget:
@ctiller https://github.com/grpc/grpc/blob/master/src/core/transport/chttp2/timeout_encoding.c#L153

**Comment by soltanmm** [comment]
Fixed in #5897 IIRC.
